File: sdk/src/logger.js
/**
 * ZYRA SHARED LOGGER
 *
 * Fire-and-forget log dispatch to the Zyra /v1/log endpoint.
 * - Non-blocking: uses setImmediate to dispatch after the current tick
 * - Graceful: never throws, never blocks the caller
 * - Respects ZYRA_DISABLE=true environment variable
 *
 * Usage:
 *   const { dispatchLog } = require('./logger')
 *   dispatchLog({ zyraKey, sdkLanguage, provider, modelRequested, ... })
 */

const https = require('https')
const http = require('http')

const ZYRA_LOG_URL = process.env.ZYRA_API_URL || 'https://api.zyra.dev'
const LOG_TIMEOUT_MS = 5000

/**
 * @typedef {Object} ZyraLogPayload
 * @property {string}  zyraKey           - Zyra API key
 * @property {string}  sdkLanguage       - 'node' | 'python' | 'proxy' | ...
 * @property {string}  provider          - 'openai' | 'anthropic' | ...
 * @property {string}  modelRequested    - model the user specified
 * @property {string}  modelRoutedTo     - model actually used
 * @property {number}  promptTokens
 * @property {number}  completionTokens
 * @property {number}  latencyMs
 * @property {number}  [costUsd]
 * @property {'success'|'error'} status
 * @property {string|null} [errorMessage]
 */

/**
 * Send a log entry to Zyra. Fire-and-forget — never awaited, never throws.
 * @param {ZyraLogPayload} payload
 */
function dispatchLog(payload) {
  if (process.env.ZYRA_DISABLE === 'true') return

  setImmediate(() => {
    try {
      _sendLog(payload)
    } catch {
      // Silently swallow — never surface to the user
    }
  })
}

function _sendLog(payload) {
  const body = JSON.stringify({
    timestamp: new Date().toISOString(),
    sdk_language: payload.sdkLanguage || 'node',
    provider: payload.provider || 'unknown',
    model_requested: payload.modelRequested || 'unknown',
    model_routed_to: payload.modelRoutedTo || payload.modelRequested || 'unknown',
    prompt_tokens: payload.promptTokens || 0,
    completion_tokens: payload.completionTokens || 0,
    latency_ms: payload.latencyMs || 0,
    cost_usd: payload.costUsd || null,
    status: payload.status || 'success',
    error_message: payload.errorMessage || null
  })

  const url = new URL(`${ZYRA_LOG_URL}/v1/log`)
  const isHttps = url.protocol === 'https:'
  const lib = isHttps ? https : http

  const options = {
    hostname: url.hostname,
    port: url.port || (isHttps ? 443 : 80),
    path: url.pathname,
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Content-Length': Buffer.byteLength(body),
      'x-zyra-api-key': payload.zyraKey || '',
      'User-Agent': `zyra-sdk-node/1.0`
    },
    timeout: LOG_TIMEOUT_MS
  }

  const req = lib.request(options)
  req.on('error', () => {})   // Silently ignore network errors
  req.on('timeout', () => { req.destroy() })
  req.write(body)
  req.end()
}

module.exports = { dispatchLog }

File: sdk-python/zyra/logger.py
"""
Zyra Python SDK — Fire-and-forget log dispatcher.

Always non-blocking. Uses a background thread so it never adds latency
to the LLM call. Gracefully swallows all errors.

Respects:
    ZYRA_DISABLE=true  — disables all logging
    ZYRA_KEY           — Zyra API key (falls back to zyra_key argument)
    ZYRA_API_URL       — override API endpoint (default: https://api.zyra.dev)
"""

import os
import json
import threading
import urllib.request
import urllib.error
from datetime import datetime, timezone
from typing import Optional

_ZYRA_API_URL = os.environ.get("ZYRA_API_URL", "https://api.zyra.dev")
_LOG_ENDPOINT = f"{_ZYRA_API_URL.rstrip('/')}/v1/log"
_LOG_TIMEOUT = 5  # seconds


def dispatch_log(
    *,
    zyra_key: Optional[str] = None,
    sdk_language: str = "python",
    provider: str = "openai",
    model_requested: str = "unknown",
    model_routed_to: Optional[str] = None,
    prompt_tokens: int = 0,
    completion_tokens: int = 0,
    latency_ms: int = 0,
    cost_usd: Optional[float] = None,
    status: str = "success",
    error_message: Optional[str] = None,
) -> None:
    """
    Dispatch a log entry to Zyra. Non-blocking — returns immediately.
    All errors are silently swallowed to protect the user's LLM call.

    Args:
        zyra_key:          Zyra API key (falls back to ZYRA_KEY env var)
        sdk_language:      'python' | 'node' | 'go' | 'proxy' | etc.
        provider:          'openai' | 'anthropic' | 'groq' | 'gemini' | ...
        model_requested:   Model the user specified (e.g. 'gpt-4o')
        model_routed_to:   Model actually used (may differ after routing)
        prompt_tokens:     Input token count
        completion_tokens: Output token count
        latency_ms:        End-to-end latency in milliseconds
        cost_usd:          Cost in USD (Zyra will estimate if None)
        status:            'success' | 'error'
        error_message:     Error message if status == 'error'
    """
    if os.environ.get("ZYRA_DISABLE") == "true":
        return

    key = zyra_key or os.environ.get("ZYRA_KEY", "")
    if not key:
        return

    payload = {
        "timestamp": datetime.now(timezone.utc).isoformat(),
        "sdk_language": sdk_language,
        "provider": provider,
        "model_requested": model_requested,
        "model_routed_to": model_routed_to or model_requested,
        "prompt_tokens": prompt_tokens,
        "completion_tokens": completion_tokens,
        "latency_ms": latency_ms,
        "cost_usd": cost_usd,
        "status": status,
        "error_message": error_message,
    }

    # Fire-and-forget via daemon thread — never blocks the caller
    t = threading.Thread(target=_send, args=(key, payload), daemon=True)
    t.start()


def _send(key: str, payload: dict) -> None:
    """Internal: performs the HTTP POST. Must not raise."""
    try:
        body = json.dumps(payload).encode("utf-8")
        req = urllib.request.Request(
            _LOG_ENDPOINT,
            data=body,
            headers={
                "Content-Type": "application/json",
                "x-zyra-api-key": key,
                "User-Agent": "zyra-sdk-python/0.2.0",
            },
            method="POST",
        )
        with urllib.request.urlopen(req, timeout=_LOG_TIMEOUT):
            pass  # We only care that it was accepted (202)
    except Exception:
        pass  # Silently swallow — never surface to user
